hawkular-alerts-engine-extensions/hawkular-alerts-aggregated-events-alerter/src/main/java/org/hawkular/alerts/extensions/ExpressionManager.java

```diff
@@ -0,0 +1,181 @@
+
+                Collection<FullTrigger> activeTriggers = new ArrayList<>();
+                for (Trigger trigger : triggers) {
+                    Collection<Condition> conditions = null;
+                    List<Condition> activeConditions = new ArrayList<>();
+                    try {
+                        if (trigger.isEnabled()) {
+                            conditions = definitions.getTriggerConditions(trigger.getTenantId(), trigger.getId(), null);
+                            log.infof("Checking [%s] Conditions for enabled trigger [%s]!", conditions.size(),
+                                    trigger.getName());
+                        }
+                    } catch (Exception e) {
+                        log.error("Failed to fetch Conditions when scheduling metrics conditions for " + trigger, e);
+                        continue;
+                    }
+                    if (null == conditions) {
+                        continue;
+                    }
+                    for (Condition condition : conditions) {
+                        if (condition instanceof ExternalCondition) {
+                            ExternalCondition externalCondition = (ExternalCondition) condition;
+                            if (TAG_VALUE.equals(externalCondition.getAlerterId())) {
+                                activeConditions.add(externalCondition);
+                            }
+                        }
+                    }
+                    if (!activeConditions.isEmpty()) {
+                        FullTrigger activeTrigger = new FullTrigger();
+                        activeTrigger.setTrigger(trigger);
+                        activeTrigger.setConditions(activeConditions);
+                        activeTriggers.add(activeTrigger);
+                    }
+                }
+                if (!activeTriggers.isEmpty()) {
+                    log.infof("ActiveTriggers: %s", activeTriggers);
+                    cep.updateConditions(activeTriggers);
+                }
+            } catch (Exception e) {
+                log.error("Failed to fetch Triggers for external conditions.", e);
+            }
+        });
+    }
+
+    @Lock(LockType.READ)
+    @Override
+    public TreeSet<Event> processEvents(TreeSet<Event> events) {
+        if (isEmpty(events)) {
+            return events;
+        }
+        List<Event> retained = new ArrayList<>();
+        TreeSet<Event> filtered = new TreeSet<>();
+        for (Event event : events) {
+            if (event.getTags().get(TAG_NAME) != null && event.getTags().get(TAG_NAME).equals(TAG_VALUE)
+                    && event.getContext().get(CONTEXT_PROCESSED) == null) {
+                retained.add(event);
+            } else {
+                filtered.add(event);
+            }
+        }
+        if (!retained.isEmpty()) {
+            executor.submit(() -> cep.processEvents(retained));
+        }
+        return filtered;
+    }
```

Comment:
Well, perhaps the javadoc is pretty basic and needs more description.

Javadoc stands:
```
    /**
     * Process Events received in the engine before it is processed by the Alerting Engine.
     *
     * @param events the Events received by the engine before to be processed
     * @return the resulting Events processed by the extension and sent to the engine
     */
    TreeSet<Event> processEvents(TreeSet<Event> events);
```

So, technically it is correct (from the generic interface perspective). 
The nuances come for this specific extension implementation: it filters a type of events from the input, so the filtered events are correct.

What is missing in the implementation is the description that the filtered events of the extension are used in an inference process and the outcome of that computation might generate additional events that it will be forwarded into the engine again.

I guess the filtered events terminology is ok here (at least for me) but I agree that without this explanation the whole process might be confusing, I will update the comments to make it better readable.